File: Algorithm/python/BAEKJOON/CLASS 2/1654.py
K, N = map(int, input().split())
K_length_list = []
for _ in range(K):
    K_length = int(input())
    K_length_list += [K_length]

min_length, max_length = 1, max(K_length_list)

while min_length <= max_length: # max _length가 min_length보다 큰 경우 다음 조건문을 실행
    mid_length = (min_length + max_length) // 2
    line_count = 0
    for length in K_length_list:
        line_count += length // mid_length

    if line_count >= N: # line_count가 N보다 크거나 같으면
        min_length = mid_length + 1 # min_length의 값을 mid_length + 1로 갱신하며 다시 mid_length를 구한다.
    else: # line_count가 N보다 작으면
        max_length = mid_length - 1 # max_length의 값을 mid_length - 1로 갱신하며 다시 max_length를 구한다.
print(max_length)

File: Algorithm/javascript/BAEKJOON/CLASS 2/1654.js
// Binary Search 사용
function solution(input) {
  let K = input[0].split(" ")[0];
  let N = input[0].split(" ")[1];
  let kLength = [];
  for (let i = 1; i <= Number(K); i++) {
    kLength.push(Number(input[i]));
  }

  // 최소길이와 최대길이를 설정
  // 두 길이의 중간값을 사용하여 kLength을 나누며 line을 Count

  let minLength = 1;
  let maxLength = Math.max(...kLength);

  let midLength = 0;
  while (minLength <= maxLength) {
    midLength = Math.floor((minLength + maxLength) / 2);
    lineCount = 0;

    for (let length of kLength) {
      lineCount += Math.floor(length / midLength);
    }

    if (lineCount >= Number(N)) {
      // lineCount가 문제에서 제시한 N보다 클 경우
      minLength = midLength + 1; // minLength를 다시 설정하여 line 자르기
    } else {
      // lineCount가 문제에서 제시한 N보다 작을 경우
      maxLength = midLength - 1; // maxLength를 다시 설정하여 line 자르기
    }
  }
  console.log(maxLength);
}

const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
rl.on("line", (line) => {
  if (!line) {
    rl.close();
  } else {
    input.push(line);
  }
}).on("close", () => {
  solution(input);
  process.exit();
});
